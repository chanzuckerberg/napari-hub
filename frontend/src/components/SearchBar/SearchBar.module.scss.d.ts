export type Styles = {
  closeIcon: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
