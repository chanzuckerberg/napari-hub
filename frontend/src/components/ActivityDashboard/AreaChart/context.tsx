import { createContext, ReactNode, useContext, useMemo } from 'react';

interface AreaChartStateValue {
  publicReleaseX: number;
  isScreen600: boolean;
  visibleMonths: number[];
}

const AreaChartStateContext = createContext<AreaChartStateValue>({
  publicReleaseX: 0,
  isScreen600: false,
  visibleMonths: [],
});

interface Props extends AreaChartStateValue {
  children: ReactNode;
}

/**
 * Provider for sharing state used in the area chart component.
 */
export function AreaChartStateProvider({
  children,
  isScreen600,
  publicReleaseX,
  visibleMonths,
}: Props) {
  const value = useMemo<AreaChartStateValue>(
    () => ({
      publicReleaseX,
      isScreen600,
      visibleMonths,
    }),
    [isScreen600, publicReleaseX, visibleMonths],
  );

  return (
    <AreaChartStateContext.Provider value={value}>
      {children}
    </AreaChartStateContext.Provider>
  );
}

export function useAreaChartState(): AreaChartStateValue {
  return useContext(AreaChartStateContext);
}
