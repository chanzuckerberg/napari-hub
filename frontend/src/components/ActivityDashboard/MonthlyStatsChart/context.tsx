import dayjs from 'dayjs';
import { createContext, ReactNode, useContext, useMemo } from 'react';

import { I18nKeys } from '@/types/i18n';

interface MonthlyStatsValue {
  dateLineI18nKey: I18nKeys<'activity'>;
  startDate: dayjs.ConfigType;
}

const MonthlyStatsContext = createContext<MonthlyStatsValue | null>(null);

interface Props
  extends Pick<MonthlyStatsValue, 'startDate' | 'dateLineI18nKey'> {
  children: ReactNode;
}

export function MonthlyStatsProvider({
  children,
  dateLineI18nKey,
  startDate,
}: Props) {
  const value = useMemo(
    () => ({
      dateLineI18nKey,
      startDate,
    }),
    [dateLineI18nKey, startDate],
  );

  return (
    <MonthlyStatsContext.Provider value={value}>
      {children}
    </MonthlyStatsContext.Provider>
  );
}

export function useMonthlyStats(): MonthlyStatsValue {
  const value = useContext(MonthlyStatsContext);

  if (!value) {
    throw new Error('useMonthlyStats() must be used in a MonthlyStatsProvider');
  }

  return value;
}
