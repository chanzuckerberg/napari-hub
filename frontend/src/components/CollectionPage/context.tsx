import { createContext, ReactNode, useContext } from 'react';

import { CollectionData } from '@/types/collections';

const CollectionContext = createContext<CollectionData | null>(null);

interface Props {
  collection: CollectionData;
  children: ReactNode;
}

/**
 * Provider for access to collection state from anywhere in collection page.
 */
export function CollectionContextProvider({ collection, children }: Props) {
  return (
    <CollectionContext.Provider value={collection}>
      {children}
    </CollectionContext.Provider>
  );
}

/**
 * Hook for accessing collection data from hub API.
 */
export function useCollection(): CollectionData {
  const collection = useContext(CollectionContext);

  if (!collection) {
    throw new Error(
      'useCollections() may not be used outside of CollectionContextProvider',
    );
  }

  return collection;
}
