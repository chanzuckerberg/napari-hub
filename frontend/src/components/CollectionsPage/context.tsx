import { createContext, ReactNode, useContext } from 'react';

import { CollectionIndexData } from '@/types/collections';

const CollectionsContext = createContext<CollectionIndexData[]>([]);

interface Props {
  collections: CollectionIndexData[];
  children: ReactNode;
}

/**
 * Provider for access to collections state from anywhere in collections home
 * page UI.
 */
export function CollectionsContextProvider({ collections, children }: Props) {
  return (
    <CollectionsContext.Provider value={collections}>
      {children}
    </CollectionsContext.Provider>
  );
}

/**
 * Hook for accessing list of collections from hub API.
 */
export function useCollections(): CollectionIndexData[] {
  return useContext(CollectionsContext);
}
