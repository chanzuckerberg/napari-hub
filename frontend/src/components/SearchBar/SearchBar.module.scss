.closeIcon path {
  stroke: theme('colors.black');
}
