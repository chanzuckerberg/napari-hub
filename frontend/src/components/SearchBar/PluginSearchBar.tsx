import { useRouter } from 'next/router';
import { HTMLProps, InputHTMLAttributes } from 'react';
import { useSnapshot } from 'valtio';

import { BEGINNING_PAGE } from '@/constants/search';
import { resetLoadingState } from '@/store/loading';
import {
  DEFAULT_SORT_TYPE,
  SEARCH_PAGE,
  SearchQueryParams,
  SearchSortType,
} from '@/store/search/constants';
import { useSearchStore } from '@/store/search/context';
import { isSearchPage, scrollToSearchBar } from '@/utils';

import { SearchBar } from './SearchBar';

interface Props extends HTMLProps<HTMLFormElement> {
  /**
   * Render large variant of search bar with a larger font size and search icon.
   */
  large?: boolean;

  /**
   * Additional props to pass to the input element.
   */
  inputProps?: InputHTMLAttributes<HTMLInputElement>;
}

/**
 * Search bar component. This renders an input field with a underline and
 * magnifying glass icon to the right of the component. When the user enters a query,
 * one of two things can happen:
 *
 * 1. User is on search page, so submitting a query re-renders the plugin list.
 *
 * 2. User is not on search page, so submitting a query redirects to the search
 * page with the `search=` URL parameter set.
 *
 * This makes the SearchBar component re-useable for non-search enabled pages.
 */
export function PluginSearchBar({ large, inputProps, ...props }: Props) {
  const router = useRouter();
  const { searchStore } = useSearchStore();
  const state = useSnapshot(searchStore);
  const { query } = state.search;

  /**
   * Performs a search query on form submission. If the user is on the search
   * page, this runs the query through the search engine. Otherwise, it
   * redirects to the search page with the query added to the URL.
   */
  async function submitForm(searchQuery: string) {
    // Reset loading state when navigating to the search page.
    resetLoadingState();

    if (isSearchPage(window.location.pathname)) {
      if (searchQuery) {
        scrollToSearchBar({ behavior: 'smooth' });
        searchStore.search.query = searchQuery;
        searchStore.sort = SearchSortType.Relevance;
      } else {
        searchStore.search.query = '';

        if (state.sort === SearchSortType.Relevance) {
          searchStore.sort = DEFAULT_SORT_TYPE;
        }
      }
    } else if (searchQuery) {
      // Set query in global state so that the search bar shows the query while
      // the search page is loading.
      searchStore.search.query = searchQuery;

      const url = {
        pathname: SEARCH_PAGE,
        query: {
          // Params will be encoded automatically by Next.js.
          [SearchQueryParams.Search]: searchQuery,
          [SearchQueryParams.Sort]: SearchSortType.Relevance,
        },
      };
      await router.push(url);
    }

    // Reset pagination when searching with a new query.
    searchStore.page = BEGINNING_PAGE;
  }

  return (
    <SearchBar
      changeOnSubmit
      value={query}
      onChange={(value) => {
        searchStore.search.query = value;
      }}
      onSubmit={(value) => submitForm(value)}
    />
  );
}
