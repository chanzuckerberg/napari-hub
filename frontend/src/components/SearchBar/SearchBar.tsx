import IconButton from '@material-ui/core/IconButton';
import clsx from 'clsx';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import { HTMLProps, useEffect, useState } from 'react';
import { useSnapshot } from 'valtio';

import { Close, Search } from '@/components/icons';
import { BEGINNING_PAGE } from '@/constants/search';
import { resetLoadingState } from '@/store/loading';
import {
  DEFAULT_SORT_TYPE,
  SEARCH_PAGE,
  SearchQueryParams,
  SearchSortType,
} from '@/store/search/constants';
import { useSearchStore } from '@/store/search/context';
import { createUrl, isSearchPage, scrollToSearchBar } from '@/utils';

import styles from './SearchBar.module.scss';

interface Props extends HTMLProps<HTMLFormElement> {
  /**
   * Render large variant of search bar with a larger font size and search icon.
   */
  large?: boolean;
}

/**
 * Search bar component. This renders an input field with a underline and
 * magnifying glass icon to the right of the component. When the user enters a query,
 * one of two things can happen:
 *
 * 1. User is on search page, so submitting a query re-renders the plugin list.
 *
 * 2. User is not on search page, so submitting a query redirects to the search
 * page with the `search=` URL parameter set.
 *
 * This makes the SearchBar component re-useable for non-search enabled pages.
 */
export function SearchBar({ large, ...props }: Props) {
  const [t] = useTranslation(['common']);
  const router = useRouter();
  const { searchStore } = useSearchStore();
  const state = useSnapshot(searchStore);
  const { query } = state.search;
  const currentPathname = createUrl(router.asPath).pathname;

  // Local state for query. This is used to store the current entered query string.
  const [localQuery, setLocalQuery] = useState(query ?? '');

  const iconClassName = clsx(
    'h-5 w-5',

    // 22x22 pixels when root font-size is 16px
    large && 'h-[1.375rem] w-[1.375rem]',
  );

  // Keep local query state in sync with global query state in case it's changed elsewhere.
  useEffect(() => {
    setLocalQuery(query);
  }, [query]);

  /**
   * Performs a search query on form submission. If the user is on the search
   * page, this runs the query through the search engine. Otherwise, it
   * redirects to the search page with the query added to the URL.
   */
  async function submitForm(searchQuery: string) {
    // Reset loading state when navigating to the search page.
    resetLoadingState();

    if (isSearchPage(window.location.pathname)) {
      if (searchQuery) {
        scrollToSearchBar({ behavior: 'smooth' });
        searchStore.search.query = searchQuery;
        searchStore.sort = SearchSortType.Relevance;
      } else {
        searchStore.search.query = '';

        if (state.sort === SearchSortType.Relevance) {
          searchStore.sort = DEFAULT_SORT_TYPE;
        }
      }
    } else if (searchQuery) {
      // Set query in global state so that the search bar shows the query while
      // the search page is loading.
      searchStore.search.query = searchQuery;

      const url = {
        pathname: SEARCH_PAGE,
        query: {
          // Params will be encoded automatically by Next.js.
          [SearchQueryParams.Search]: searchQuery,
          [SearchQueryParams.Sort]: SearchSortType.Relevance,
        },
      };
      await router.push(url);
    }

    // Reset pagination when searching with a new query.
    searchStore.page = BEGINNING_PAGE;
  }

  return (
    <form
      data-testid="searchBarForm"
      className={clsx(
        // Flex layout
        'flex flex-auto items-center',

        // Borders
        'border-b-2 border-black',

        large && 'text-xl',
      )}
      onSubmit={async (event) => {
        event.preventDefault();
        await submitForm(localQuery);
      }}
      {...props}
    >
      <input
        aria-label={t('common:ariaLabels.searchBar')}
        data-testid="searchBarInput"
        className={clsx(
          // Flex layout
          'flex flex-auto',

          // Remove border and focus outline around input
          'border-none outline-none',

          // Remove white colored input background
          'bg-transparent',

          /*
            Inputs have a default width defined by the browser, so we have to
            set this explicitly to make the input flexible:
            https://stackoverflow.com/a/42421490
          */
          'w-0',
        )}
        onChange={(event) => {
          const { value } = event.target;
          setLocalQuery(value);
        }}
        value={localQuery}
      />

      <IconButton
        aria-label={t(
          query
            ? 'common:ariaLabels.clearSearchBar'
            : 'common:ariaLabels.submitSearchQuery',
        )}
        data-testid={query ? 'clearQueryButton' : 'submitQueryButton'}
        onClick={async () => {
          // Clear local query if close button is clicked and the search engine
          // is currently rendering the results for another query.
          let searchQuery = localQuery;

          if (query) {
            setLocalQuery('');
            searchQuery = '';
          }

          await submitForm(searchQuery);
        }}
      >
        {/* Render close button if the user submitted a query. */}
        {query && isSearchPage(currentPathname) ? (
          <Close className={clsx(iconClassName, styles.closeIcon)} />
        ) : (
          <Search className={iconClassName} />
        )}
      </IconButton>
    </form>
  );
}
