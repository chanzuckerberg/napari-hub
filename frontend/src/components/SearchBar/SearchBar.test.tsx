import { render } from '@testing-library/react';

import { SearchStoreProvider } from '@/store/search/context';

import { SearchBar } from './SearchBar';

describe('<SearchBar />', () => {
  it('should match snapshot', () => {
    const component = render(
      <SearchStoreProvider>
        <SearchBar />
      </SearchStoreProvider>,
    );
    expect(component.asFragment()).toMatchSnapshot();
  });
});
