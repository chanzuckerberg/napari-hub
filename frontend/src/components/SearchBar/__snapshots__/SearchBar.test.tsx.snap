// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-auto items-center border-b-2 border-black"
    data-testid="searchBarForm"
  >
    <input
      aria-label="Search bar for searching for napari plugins."
      class="flex flex-auto border-none outline-none bg-transparent w-0"
      data-testid="searchBarInput"
      value=""
    />
    <button
      aria-label="Submit search query"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
      data-testid="submitQueryButton"
      tabindex="0"
      type="button"
    >
      <svg
        class="h-5 w-5"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.32004 7.68001L0.746704 15.2533M10.6667 9.60001C13.0134 9.60001 14.9334 7.68001 14.9334 5.33335C14.9334 2.98668 13.0134 1.06668 10.6667 1.06668C8.32004 1.06668 6.40004 2.98668 6.40004 5.33335C6.40004 7.68001 8.32004 9.60001 10.6667 9.60001Z"
          stroke="black"
          stroke-width="2"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
